File: crates/faf-app/src/services/mods.rs
//! Mods service.
//!
//! Thin handler (like `services/maps.rs`): asks the [`crate::ports::
//! ModsPort`] to do the work, then emits the corresponding events. The
//! actual API calls, folder scan, zip extraction, and `game.prefs`
//! read/write live entirely behind the port: see `infra/mods.rs`.

use faf_domain::state::{ModListStatus, ModsCommand, ModsEvent};

use crate::runtime::{EventSink, ServiceCtx};

pub async fn handle(cmd: ModsCommand, ctx: &ServiceCtx, out: &EventSink) {
    match cmd {
        ModsCommand::LoadVault => {
            // Same guard, same reason, as `services::maps`: one crawl.
            if out.with_state(|state| {
                matches!(
                    state.mods.vault_status,
                    ModListStatus::Loading | ModListStatus::Ready
                )
            }) {
                return;
            }
            out.emit(ModsEvent::VaultLoading);
            match ctx.ports.mods.list_vault().await {
                Ok(mods) => out.emit(ModsEvent::VaultLoaded { mods }),
                Err(reason) => out.emit(ModsEvent::VaultLoadFailed { reason }),
            }
        }
        ModsCommand::SearchVault { query } => {
            out.emit(ModsEvent::VaultSearching);
            match ctx.ports.mods.search_vault(query.clone()).await {
                Ok(page) => out.emit(ModsEvent::VaultSearched {
                    mods: page.mods,
                    query,
                    total_pages: page.total_pages,
                    total_records: page.total_records,
                }),
                Err(reason) => out.emit(ModsEvent::VaultSearchFailed { reason }),
            }
        }
        ModsCommand::LoadInstalled => {
            out.emit(ModsEvent::InstalledLoading);
            match ctx.ports.mods.list_installed().await {
                Ok(mods) => out.emit(ModsEvent::InstalledLoaded { mods }),
                Err(reason) => out.emit(ModsEvent::InstalledLoadFailed { reason }),
            }
        }
        ModsCommand::InstallMod { uid, download_url } => {
            let _guard = ctx.mods_mutation.acquire().await;
            out.emit(ModsEvent::Installing { uid: uid.clone() });
            match ctx.ports.mods.install_mod(uid, download_url).await {
                Ok(installed) => out.emit(ModsEvent::Installed { installed }),
                Err(reason) => out.emit(ModsEvent::InstallFailed { reason }),
            }
        }
        ModsCommand::UninstallMod { folder_name, uid } => {
            let _guard = ctx.mods_mutation.acquire().await;
            out.emit(ModsEvent::Installing { uid });
            match ctx.ports.mods.uninstall_mod(folder_name).await {
                Ok(installed) => out.emit(ModsEvent::Uninstalled { installed }),
                Err(reason) => out.emit(ModsEvent::UninstallFailed { reason }),
            }
        }
        ModsCommand::ToggleMod { uid, enabled } => {
            let _guard = ctx.mods_mutation.acquire().await;
            out.emit(ModsEvent::Toggling { uid: uid.clone() });
            match ctx.ports.mods.toggle_mod(uid, enabled).await {
                Ok(installed) => out.emit(ModsEvent::Toggled { installed }),
                Err(reason) => out.emit(ModsEvent::ToggleFailed { reason }),
            }
        }
    }
}

File: ui/src/store/reducers/mods.ts
import type { ModsEvent, ModsState } from "../../ipc/bindings";

export function reduceMods(state: ModsState, event: ModsEvent): ModsState {
  switch (event.type) {
    case "vaultLoading":
      return { ...state, vaultStatus: { type: "loading" } };
    case "vaultLoaded":
      return { ...state, vault: event.payload.mods, vaultStatus: { type: "ready" } };
    case "vaultLoadFailed":
      return { ...state, vaultStatus: { type: "failed", payload: { reason: event.payload.reason } } };
    case "vaultSearching":
      return { ...state, browseStatus: { type: "loading" } };
    case "vaultSearched":
      return {
        ...state,
        browse: event.payload.mods,
        browseQuery: event.payload.query,
        browseTotalPages: event.payload.totalPages,
        browseTotalRecords: event.payload.totalRecords,
        browseStatus: { type: "ready" },
      };
    case "vaultSearchFailed":
      return {
        ...state,
        browseStatus: { type: "failed", payload: { reason: event.payload.reason } },
      };
    case "installedLoading":
      return { ...state, installedStatus: { type: "loading" } };
    case "installedLoaded":
      return { ...state, installed: event.payload.mods, installedStatus: { type: "ready" } };
    case "installedLoadFailed":
      return {
        ...state,
        installedStatus: { type: "failed", payload: { reason: event.payload.reason } },
      };
    case "installing":
      return {
        ...state,
        installStatus: { type: "installing", payload: { uid: event.payload.uid } },
      };
    case "installed":
      return {
        ...state,
        installed: event.payload.installed,
        installedStatus: { type: "ready" },
        installStatus: { type: "idle" },
      };
    case "installFailed":
      return { ...state, installStatus: { type: "failed", payload: { reason: event.payload.reason } } };
    case "uninstalled":
      return {
        ...state,
        installed: event.payload.installed,
        installedStatus: { type: "ready" },
        installStatus: { type: "idle" },
      };
    case "uninstallFailed":
      return { ...state, installStatus: { type: "failed", payload: { reason: event.payload.reason } } };
    case "toggling":
      return {
        ...state,
        toggleStatus: { type: "toggling", payload: { uid: event.payload.uid } },
      };
    case "toggled":
      return {
        ...state,
        installed: event.payload.installed,
        installedStatus: { type: "ready" },
        toggleStatus: { type: "idle" },
      };
    case "toggleFailed":
      return { ...state, toggleStatus: { type: "failed", payload: { reason: event.payload.reason } } };
    case "downloadSizesResolved": {
      // Only the answers that carried a number. A uid that is absent has not
      // been asked about or came back without a length, and both mean "do not
      // print a size", which is the same rule the Rust reducer applies.
      const sizes = { ...state.downloadSizes };
      for (const size of event.payload.sizes) {
        if (size.bytes !== null) sizes[size.uid] = size.bytes;
      }
      return { ...state, downloadSizes: sizes };
    }
  }
}
